File: Data Structures & Algorithms/binary-tree-diameter/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    int ans = 0;

    public int solve(TreeNode root){
        if(root==null){
            return 0;
        }

        int left = solve(root.left);
        int right = solve(root.right);

        int currAns = 1 + Math.max(left, right);
        ans = Math.max(ans, right + left);

        return currAns;
    }
    public int diameterOfBinaryTree(TreeNode root) {
        solve(root);
        return ans;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.cpp
class Solution {
public:

    int characterReplacement(string s, int k) {

        unordered_map<char, int> mp;
        int n = s.size();
        int ans = 0;

        int i = 0;
        int j = 0;

        int maxF = 0;

        while(j<n){

            mp[s[j]]++;
            maxF = max(maxF, mp[s[j]]);

            while((j-i+1) - maxF > k){
                mp[s[i]]--;
                i++;
            }

            ans = max(ans, j-i+1);
            j++;
            
        }

        return ans;


    }
};
